File: Baekjoon/11779 - 최소비용 구하기 2/11779.cpp
#include <iostream>
#include <algorithm>
#include <vector>
using namespace std;

#define FASTIO cin.tie(NULL); cout.tie(NULL); ios::sync_with_stdio(false);
#define INF 1000000000

int main() {
    FASTIO
    
    int n, m, a, b;

    cin >> n;
    cin >> m;

    vector<vector<int>> bus(n+1, vector<int>(n+1, INF));
    vector<int> cost(n+1, INF);
    vector<bool> visit(n+1, false);
    vector<vector<int>> route(n+1, vector<int>(0));

    for(int _ = 0; _ < m; _++) {
        int s, e, c;
        cin >> s >> e >> c;

        bus[s][e] = min(bus[s][e], c);
    }

    cin >> a >> b;

    for(int i = 1; i <= n; i++) {
        bus[i][i] = 0;
        cost[i] = bus[a][i];

        route[i].push_back(a);
        if(i != a) route[i].push_back(i);
    }

    visit[a] = true;

    for(int _ = 0; _ < n-2; _++) {
        int small = INF, next = 0;
        
        for(int i = 1; i <= n; i++) {
            if(visit[i] == false && cost[i] <= small) {
                small = cost[i];
                next = i;
            }
        }

        visit[next] = true;

        for(int i = 1; i <= n; i++) {
            if(visit[i] == false) {
                if(cost[next] + bus[next][i] < cost[i]) {
                    cost[i] = cost[next] + bus[next][i];

                    route[i].clear();
                    for(int j = 0; j < route[next].size(); j++) route[i].push_back(route[next][j]);
                    route[i].push_back(i);
                }
            }
        }
    }

    cout << cost[b] << "\n";
    cout << route[b].size() << "\n";
    for(auto routes : route[b]) cout << routes << " ";
}

File: Baekjoon/11779 - 최소비용 구하기 2/11779.py
import sys
input = sys.stdin.readline

INF = 1e9

n = int(input())
m = int(input())

bus = [[INF for _ in range(n+1)] for _ in range(n+1)]
cost = [INF for _ in range(n+1)]
visit = [False for _ in range(n+1)]
route = [list() for _ in range(n+1)]

for _ in range(m):
    s, e, c = map(int, input().split())
    bus[s][e] = min(bus[s][e], c)

a, b = map(int, input().split())

for i in range(1, n+1):
    bus[i][i] = 0
    cost[i] = bus[a][i]

    route[i].append(a)
    if i != a:
        route[i].append(i)

visit[a] = True

for _ in range(n-2):
    small = INF
    next = 0

    for i in range(1, n+1):
        if visit[i] == False and cost[i] <= small:
            small = cost[i]
            next = i
    
    visit[next] = True

    for i in range(1, n+1):
        if visit[i] == False:
            if cost[next] + bus[next][i] < cost[i]:
                cost[i] = cost[next] + bus[next][i]

                route[i].clear()
                for routes in route[next]:
                    route[i].append(routes)
                route[i].append(i)

print(cost[b])
print(len(route[b]))
for routes in route[b]:
    print(routes, end=" ")